test/parallel/test-runner-bail.js

```diff
@@ -0,0 +1,75 @@
+'use strict';
+
+require('../common');
+const fixtures = require('../common/fixtures');
+const tmpdir = require('../common/tmpdir');
+const { describe, it } = require('node:test');
+const { spawnSync } = require('node:child_process');
+const assert = require('node:assert');
+
+const testFile = fixtures.path('test-runner/bail/bail.js');
+const errorTestFile = fixtures.path('test-runner/bail/error.js');
+
+tmpdir.refresh();
+
+describe('maintain errors', () => {
+  it('should exit at assertion failure', () => {
+    const child = spawnSync(process.execPath, ['--test', '--test-bail', errorTestFile]);
+    assert.strictEqual(child.stderr.toString(), '');
+    assert.match(child.stdout.toString(), /failureType: 'testCodeFailure'/);
+    assert.match(child.stdout.toString(), /0 !== 1/);
+    assert.match(child.stdout.toString(), /AssertionError/);
+    assert.doesNotMatch(child.stdout.toString(), /not ok 2 - dont show/);
+    assert.doesNotMatch(child.stdout.toString(), /# Subtest: dont show/);
+  });
+});
+
+describe('node:test bail tap', () => {
```

Comment:
TAP actually has its own syntax for bailing out (see https://testanything.org/tap-specification.html). It's not strictly necessary, but it would be nice if we could leverage it.

Reply:
So we should print `Bail out!` + a fixed message once it bails out

Reply:
I don't think we need to do that.

The bailing out action is handled by all of the tests being cancelled. We could emit a single `node:bail` event on the reporter when the bailout process starts. The TAP reporter could detect this event and print the `Bail out!` message (and potentially choose to exit the process - I'm not sure what is ideal there). Any other reporter could treat the event however they want.

Reply:
this would look ok?
<img width="1111" alt="Screenshot 2023-08-14 at 12 27 59" src="https://github.com/nodejs/node/assets/36735501/a3ef852d-13cb-4ed7-b2da-227a17b5e76d">
I took inspiration from
https://node-tap.org/tap-protocol/#bail-out but I'm not 100%

Reply:
I actally decided not too implement this at the moment and keep it for a future pr